pkg/location/parse_test.go

```diff
@@ -0,0 +1,185 @@
+				Directory: ".",
+			},
+			wantErr: false,
+		},
+		{
+			name: "OciNameWithTag",
+			args: args{
+				location: "oci://my-registry.local/name:tag",
+			},
+			want: Oci{
+				Image:     name.MustParseReference("my-registry.local/name:tag"),
+				Directory: ".",
+			},
+			wantErr: false,
+		},
+		{
+			name: "OciNameWithDigest",
+			args: args{
+				location: "oci://my-registry.local/name@sha256:9f6ca9562c5e7bd8bb53d736a2869adc27529eb202996dfefb804ec2c95237ba",
+			},
+			want: Oci{
+				Image:     name.MustParseReference("my-registry.local/name@sha256:9f6ca9562c5e7bd8bb53d736a2869adc27529eb202996dfefb804ec2c95237ba"),
+				Directory: ".",
+			},
+			wantErr: false,
+		},
+		{
+			name: "OciNameWithDirectory",
+			args: args{
+				location: "oci://my-registry.local/name//sub/directory",
+			},
+			want: Oci{
+				Image:     name.MustParseReference("my-registry.local/name:latest"),
+				Directory: "sub/directory",
+			},
+			wantErr: false,
+		},
+		{
+			name: "OciNameWithDirectoryAndTag",
+			args: args{
+				location: "oci://my-registry.local/name//sub/directory:tag",
+			},
+			want: Oci{
+				Image:     name.MustParseReference("my-registry.local/name:tag"),
+				Directory: "sub/directory",
+			},
+			wantErr: false,
+		},
+		{
+			name: "OciDirectoryExtraSlashes",
+			args: args{
+				location: "oci://my-registry.local/name///sub/directory",
+			},
+			want: Oci{
+				Image:     name.MustParseReference("my-registry.local/name:latest"),
+				Directory: "sub/directory",
+			},
+			wantErr: false,
+		},
+		{
+			name: "OciDirectoryEmptyPath",
+			args: args{
+				location: "oci://my-registry.local/name//",
+			},
+			want: Oci{
+				Image:     name.MustParseReference("my-registry.local/name:latest"),
+				Directory: ".",
+			},
+			wantErr: false,
+		},
+		{
+			name: "GitSimpleRepo",
+			args: args{
+				location: "https://hostname/repo.git@main",
+			},
+			want: Git{
+				Repo:      "https://hostname/repo",
+				Directory: ".",
+				Ref:       "main",
+			},
+			wantErr: false,
+		},
+		{
+			name: "GitWithDirectory",
+			args: args{
+				location: "https://hostname/repo.git/sub/directory@main",
+			},
+			want: Git{
+				Repo:      "https://hostname/repo",
+				Directory: "sub/directory",
+				Ref:       "main",
+			},
+			wantErr: false,
+		},
+		{
+			name: "SimpleDir",
+			args: args{
+				location: "path/to/directory",
+			},
+			want: Dir{
+				Directory: "path/to/directory",
+			},
+			wantErr: false,
+		},
+		{
+			name: "InputStream",
+			args: args{
+				location: "-",
+				opts:     []Option{WithStdin(testReader)},
+			},
+			want: InputStream{
+				Reader: testReader,
+			},
+			wantErr: false,
+		},
+		{
+			name: "OutputStream",
+			args: args{
+				location: "-",
+				opts:     []Option{WithStdout(testWriter)},
+			},
+			want: OutputStream{
+				Writer: testWriter,
+			},
+			wantErr: false,
+		},
+		{
+			name: "StreamLocationNotExpected",
+			args: args{
+				location: "-",
+			},
+			want: Dir{
+				Directory: "-",
```

Comment:
that is true. there is an ambiguity added by "-" as stdio which is a valid dir name. 

some points that kind of mitigate the concern, 
* It's the code calling ParseReference that can decide which formats are accepted for that particular arg. So it's not a blanket collision, instead it's only where a CLI makes a choice, for example, that a `[DIR]` destination arg is actually a `[DIR] | -` filesystem or stdio arg.
* For those cases, you can point to `kubectl apply -f -` to read from stdin instead of a folder name. It doesn't avoid the ambiguity, but at least it establishes precedent.
* Finally the person calling the CLI can also use an "/absolute" or "./relative" path to guarantee it matches the `[DIR]` format only. So that works for a hardcoded `"./-"` dir if needed, or `"./${RELPATH}"` if it's in a script and the path isn't known in advance...